feature/security/gnsi/certz/tests/client_certificates/client_certificates_test.go

```diff
@@ -0,0 +1,254 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package client_certificates_test
+
+import (
+	context "context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"os"
+	"testing"
+
+	setupService "github.com/openconfig/featureprofiles/feature/security/gnsi/certz/tests/internal/setup_service"
+	"github.com/openconfig/featureprofiles/internal/fptest"
+	"github.com/openconfig/ondatra"
+	"github.com/openconfig/ondatra/gnmi"
+	"github.com/openconfig/ondatra/gnmi/oc"
+	"github.com/openconfig/ygot/ygot"
+)
+
+var (
+	testProfile = "gNxI"
+	serverAddr  string
+	username    = "certzuser"
+	password    = "certzpswd"
+)
+
+// createUser function to add an user in admin role.
+func createUser(t *testing.T, dut *ondatra.DUTDevice, user, pswd string) bool {
+	ocUser := &oc.System_Aaa_Authentication_User{
+		Username: ygot.String(user),
+		Role:     oc.AaaTypes_SYSTEM_DEFINED_ROLES_SYSTEM_ROLE_ADMIN,
+		Password: ygot.String(pswd),
+	}
+	gnmi.Update(t, dut, gnmi.OC().System().Aaa().Authentication().User(user).Config(), ocUser)
+	return true
+}
+
+func TestMain(m *testing.M) {
+	fptest.RunTests(m)
+}
+
+// TestClientCertTcOne Test that client certificates from a set of one CA are able to be validated and
+// used for authentication to a device when used by a client connecting to each
+// gRPC service.
+
+func TestClientCert(t *testing.T) {
+
+	dut := ondatra.DUT(t, "dut")
+	serverAddr = dut.Name()
+
+	checkUser := createUser(t, dut, username, password)
+	if !checkUser {
+		t.Fatalf("Failed to create certz user.")
+	}
+
+	t.Logf("Validation of all services that are using gRPC before certz rotation.")
+	initCheck := setupService.PreInitCheck(context.Background(), t, dut)
+	if !initCheck {
+		t.Fatalf("Failed in the preInit checks.")
+	}
+
+	ctx := context.Background()
+	gnsiC, err := dut.RawAPIs().BindingDUT().DialGNSI(ctx)
+	if err != nil {
+		t.Fatalf("Could not create gNSI Connection %v", err)
+	}
+	t.Logf("Precheck:gNSI connection is successful %v", gnsiC)
+
+	cases := []struct {
+		desc             string
+		serverCertzFile  string
+		serverKeyzFile   string
+		trustBundlezFile string
+		clientCertzFile  string
+		clientKeyzFile   string
+		mismatch         bool
+	}{
+		{
+			desc:             "Certz1.1:Load the correct key-type rsa trustbundle with 1 CA configuration",
+			serverCertzFile:  "test_data/ca-01/server-rsa-a-cert.pem",
```

Comment:
is the reason that a bunch of the test data/etc is copied around because ../<foo>/test_data/.... is not feasible for some reason?

I'd also tend to make 'test_data' a const that points to the correct place:
  const(
    test_data = "../foo/test_data"
  )
  
  at the top of this test file.

Reply:
noted .